src/order/countable_dense_linear_order.lean

```diff
@@ -0,0 +1,288 @@
+  left_inv := begin
+    intro a,
+    suffices : cmp a (g $ f a) = ordering.eq,
+    { symmetry, simpa using this },
+    rw adj, simp,
+  end,
+  right_inv := begin
+    intro b,
+    suffices : cmp (f $ g b) b = ordering.eq,
+    { simpa using this },
+    rw ←adj, simp,
+  end,
+  ord' := begin
+    intros x y,
+    apply gt_iff_gt_of_cmp_eq,
+    suffices : y = g (f y),
+    { convert adj x (f y), },
+    suffices : cmp y (g $ f y) = ordering.eq,
+    { simpa using this },
+    rw adj, simp,
+  end }
+
+end orderiso
+
+section swap
+
+lemma swap_eq_iff {α β} {p : α × β} {q : β × α} : p.swap = q ↔ p = q.swap :=
+begin
+  split,
+  { intro h, rw [←h, prod.swap_swap] },
+  { intro h, rw [h, prod.swap_swap] }
+end
+
+lemma mem_swap {α β} [decidable_eq α] [decidable_eq β] {s : finset $ α × β} {p : β × α} :
+  p ∈ s.image prod.swap ↔ p.swap ∈ s :=
+begin
+  rw finset.mem_image,
+  exact ⟨λ ⟨q, h, he⟩, swap_eq_iff.mp he ▸ h, λ h, ⟨p.swap, h, prod.swap_swap p⟩⟩,
+end
+
+lemma sub_swap_iff {α β} [decidable_eq α] [decidable_eq β] {s : finset $ α × β}
+  {t : finset $ β × α} : s ⊆ t.image prod.swap ↔ s.image prod.swap ⊆ t :=
+begin
+  rw [finset.subset_iff, finset.subset_iff],
+  split,
+  { intros h p hp,
+    rw mem_swap at hp,
+    specialize h hp,
+    rwa [mem_swap, prod.swap_swap] at h, },
+  { intros h p hp,
+    rw mem_swap,
+    apply h,
+    rwa [mem_swap, prod.swap_swap], }
+end
+
+end swap
+
+-- For any finsets lo, hi, with all elements of lo less than those of hi,
+-- there is an element m that separates them.
+lemma separation {α : Type*} [decidable_linear_order α]
+  [densely_ordered α] [no_bot_order α] [no_top_order α] [nonem : nonempty α]
+  (lo hi : finset α) (lo_lt_hi : ∀ (x ∈ lo) (y ∈ hi), x < y) :
+  ∃ m : α, (∀ x ∈ lo, x < m) ∧ (∀ y ∈ hi, m < y) :=
+if nlo : lo.nonempty then
+  if nhi : hi.nonempty then
+    exists.elim (dense (lo_lt_hi _ (finset.max'_mem _ nlo) _ (finset.min'_mem _ nhi))) (λ m hm,
+    ⟨m, λ x hx, lt_of_le_of_lt (finset.le_max' lo nlo x hx) hm.1,
+        λ y hy, lt_of_lt_of_le hm.2 (finset.min'_le hi nhi y hy)⟩)
+  else
+    exists.elim (no_top (finset.max' lo nlo)) (λ m hm,
+    ⟨m, λ x hx, lt_of_le_of_lt (finset.le_max' lo nlo x hx) hm,
+        λ y hy, (nhi ⟨y, hy⟩).elim⟩)
+else
+  if nhi : hi.nonempty then
+    exists.elim (no_bot (finset.min' hi nhi)) (λ m hm,
+    ⟨m, λ x hx, (nlo ⟨x, hx⟩).elim,
+        λ y hy, lt_of_lt_of_le hm (finset.min'_le hi nhi y hy)⟩)
+  else
+    nonem.elim (λ m,
+    ⟨m, λ x hx, (nlo ⟨x, hx⟩).elim,
+        λ y hy, (nhi ⟨y, hy⟩).elim⟩)
+
+variables (α : Type*) [decidable_linear_order α]
+  (β : Type*) [decidable_linear_order β]
+
+/-- The type of partial order isomorphism between `α` and `β` defined on finite subsets.
+    A partial order isomorphism is encoded as a finite subset of `α × β`, consisting
+    of pairs which should be identified. -/
+def partial_iso : Type _ :=
+{ f : finset (α × β) // ∀ (p q ∈ f),
+  cmp (prod.fst p) (prod.fst q) = cmp (prod.snd p) (prod.snd q) }
```

Comment:
There's some indirection throughout this file where you use `cmp` - is it easier to use
```lean
{ f : finset (α × β) // ∀ (p q ∈ f), prod.fst p < prod.fst q ↔ prod.snd p < prod.snd q }
```
instead? (This is equivalent since a finset has no dups)

Reply:
Not sure. The assumption `cmp a b = cmp x y` is nice because it carries all the information you'll ever need. We could also just assume `prod.fst p < prod.fst q →prod.snd p < prod.snd q`, but this isn't very symmetric.